Add Pkpass parser

TheLastProject opened the pull request:
!!! Blocked by #2051 !!!

- [X] Parse pkpass files
- [ ] Implement Stocard Zip bugfix
- [x] Load parsed images into edit activity
- [X] Basic tests
- [ ] Consider adding more tests (we got more files from users)
- [x] Integrate with scan activity
- [x] Integrate with clicking pkpass file in file manager
- [x] Don't crash when pkpassparser throws an exception
- [ ] Logo background colour takes priority over background colour defined in card (due to LoyaltyCardEditActivity setting headerColor on image thumbnail set)

New regression to fix:
- [x] Only load non-thumbnail images when actually needed
- [ ] Rotating the edit activity without changes causes Catima to think you made a change

theimpulson commented on app/src/main/java/protect/card_locker/PkpassParser.kt:
```diff
@@ -0,0 +1,404 @@
+            }
+
+            try {
+                zipInputStream.close()
```
Tip: With streams, you can use the kotlin's `.use` method to close them when done.
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/use.html

theimpulson replied:
Example: https://gitlab.com/AuroraOSS/AuroraStore/-/blob/master/app/src/main/java/com/aurora/store/data/work/DownloadWorker.kt#L294

TheLastProject replied:
Thanks, also for the pointer! It's basically Python's `with` statement, it does feel cleaner and helps scope more explicitly :)

theimpulson commented on app/src/main/java/protect/card_locker/PkpassParser.kt:
```diff
@@ -0,0 +1,404 @@
+
+    fun getImage(): Bitmap? {
+        return image
+    }
```
You can change the variable to specify a getter directly like `private val image: Bitmap get() = <load_something>`. This will behave as a function and you get the benefit of nice syntax.

TheLastProject replied:
My main concern is that I don't want the variable to be set from the outside, but I do want it readable from the outside. I guess your suggestion here is to rename the internal variable away from "image"?

TheLastProject commented:
So, I've spend way too much time on this. This whole feature has been... challenging. Every little fix broke something else and I ended up having to change a huge amount of Catima to make this work (and future-proof) but I think we might be here. [All of this to fix the lightbulb in the kitchen](https://www.youtube.com/watch?v=AbSehcT19u0).

I would be great if people could do some testing to see if I missed anything. And not just of the pkpass part, just of... the whole app, because so many core things about card handling got changed.

If nobody reports any issues soon, I'll probably just merge this and give the translators some time to translate and then release :)

[app-debug.zip](https://github.com/user-attachments/files/18011606/app-debug.zip)

(Note: this is a debug build with a different package name, it will be installed as a second "Catima Debug" app, it won't override your main app)

TheLastProject commented:
Oops, I forgot to handle the pkpass parser being unable to parse the file...

theimpulson commented on build.gradle.kts:
```diff
@@ -3,6 +3,7 @@
 plugins {
     id("com.android.application") version "8.7.3" apply false
     id("com.github.spotbugs") version "5.1.4" apply false
+    id("org.jetbrains.kotlin.android") version "2.0.0" apply false
```
```suggestion
    id("org.jetbrains.kotlin.android") version "2.1.0" apply false
```

theimpulson commented on app/src/main/java/protect/card_locker/ParseResultType.kt:
```diff
@@ -0,0 +1,6 @@
+enum class ParseResultType {
+    FULL,
+    BARCODE_ONLY
+}
```
Missing new line

theimpulson commented on app/src/main/java/protect/card_locker/ParseResult.kt:
```diff
@@ -0,0 +1,31 @@
+            }
+        }
+    }
+}
```
Missing new line

theimpulson commented on app/src/main/java/protect/card_locker/PkpassParser.kt:
```diff
@@ -0,0 +1,404 @@
+            throw IOException(context.getString(R.string.errorReadingFile))
+        }
+
+        try {
```
Consider extracting this try-and-catch logic as a separate function to improve readability.

TheLastProject commented:
New debug build:

- Fixes crashing on parsing failure
- Fixes not parsing rgb(0,0,0) colours (though the thumbnail image colour will override it, this is kind of a bug)
- Fixed some unnecesary double loading-from-storage issues
- Updated dependencies and some other @theimpulson suggested cleanups (thanks)

[app-debug.zip](https://github.com/user-attachments/files/18028800/app-debug.zip)

TheLastProject commented:
Okay, I've had multiple people tell me it works great now and all my testing seems to succeed to.

I guess it's time to hit the merge button :)

Thank you for testing, everyone! There's still some follow-up features that would be neat (like updating the pkpass card) but this PR has already taken much longer than I wanted so let's just call this done for now, give the translators a moment to translate, and do a new release.

If anyone still finds anything in the latest beta release after this message, still please *do* tell me :)
